src/algorithms/l0_smooth.jl

```diff
@@ -0,0 +1,51 @@
+struct L0Smooth <: AbstractImageSmoothAlgorithm
+    """smoothing weight"""
+    λ::Float64
+    """iteration rate"""
+    κ::Float64
+
+    function L0Smooth(λ, κ)
+        λ > zero(λ) || throw(ArgumentError("smoothing weight λ should be positive"))
+        κ > one(κ) || throw(ArgumentError("iteration rate κ should be > 1.0"))
+        new(λ, κ)
+    end
+end
+
+L0Smooth(; λ::Float64=2e-2, κ::Float64=2.0) = L0Smooth(λ, κ)
+
+function (f::L0Smooth)(out::GenericGrayImage,
+                       img::GenericGrayImage)
+    S = float.(channelview(img))
+    λ = f.λ
+    κ = f.κ
+    βmax = 1e5
+    fx = [1 -1]
+    fy = [1, -1]
+    N, M = size(S)
+    sizeI2D = (N, M)
+    sizeI2D_t = (M, N)
+    otfFx = freqkernel(centered(fx), sizeI2D)
+    otfFy = transpose(freqkernel(centered(transpose(fy)), sizeI2D_t))
+    Normin1 = fft(S, (1, 2))
+    Denormin2 = abs.(otfFx).^2 + abs.(otfFy).^2
+    β = 2*λ
+    while β < βmax
+        Denormin = 1 .+ β * Denormin2

```

Comment:
This `Denormin` memory can be pre-allocated outside this while loop and reused.

similar to `h` and `v`, and `Normin2`, where you need to implement an in-place version of `forwarddiff` and `backdiff`.